contracts/finance/VestingWalletConfidentialFactory.sol

```diff
@@ -0,0 +1,175 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import {FHE, euint64, externalEuint64, euint128, ebool} from "@fhevm/solidity/lib/FHE.sol";
+import {Clones} from "@openzeppelin/contracts/proxy/Clones.sol";
+import {IConfidentialFungibleToken} from "../interfaces/IConfidentialFungibleToken.sol";
+import {VestingWalletCliffConfidential} from "./VestingWalletCliffConfidential.sol";
+import {VestingWalletConfidential} from "./VestingWalletConfidential.sol";
+import {VestingWalletExecutorConfidential} from "./VestingWalletExecutorConfidential.sol";
+
+abstract contract VestingWalletConfidentialFactory {
+    address private immutable _vestingWalletConfidentialImplementation;
+
+    error VestingWalletConfidentialInvalidDuration();
+    error VestingWalletConfidentialInvalidStartTimestamp(address beneficiary, uint64 startTimestamp);
+
+    /**
+     * @dev
+     */
+    event VestingWalletConfidentialBatchFunded(ebool success);
+    /**
+     * @dev
+     */
+    event VestingWalletConfidentialCreated(
+        address indexed beneficiary,
+        address indexed vestingWalletConfidential,
+        uint48 startTimestamp
+    );
+
+    /**
+     * @dev
+     */
+    struct VestingPlan {
+        address beneficiary;
+        externalEuint64 encryptedAmount;
+        uint48 startTimestamp;
+        uint48 cliff;
+        address executor;
+    }
+
+    /**
+     * @dev
+     */
+    constructor() {
+        _vestingWalletConfidentialImplementation = address(new VestingWalletCliffExecutorConfidential());
+    }
+
+    /**
+     * @dev Batches the funding of multiple confidential vesting wallets.
+     *
+     * Funds are sent to predeterministic wallet addresses. Wallets can be created later.
+     */
+    function batchFundVestingWalletConfidential(
+        address confidentialFungibleToken,
+        externalEuint64 totalEncryptedAmount,
+        bytes calldata inputProof,
+        VestingPlan[] calldata vestingPlans,
+        uint48 durationSeconds
+    ) external returns (ebool) {
+        require(durationSeconds > 0, VestingWalletConfidentialInvalidDuration());
+        euint64 totalTransferedAmount = euint64.wrap(0);
+        uint256 vestingPlansLength = vestingPlans.length;
+        for (uint256 i = 0; i < vestingPlansLength; i++) {
+            VestingPlan memory vestingPlan = vestingPlans[i];
+            euint64 encryptedAmount = FHE.fromExternal(vestingPlan.encryptedAmount, inputProof);
+            require(
+                vestingPlan.startTimestamp >= block.timestamp,
+                VestingWalletConfidentialInvalidStartTimestamp(vestingPlan.beneficiary, vestingPlan.startTimestamp)
```

Comment:
We should just duplicate the restrictions enforced on the vesting wallet side. I think it's ok if the vesting start timestamp is in the past.

Reply:
I don't see any restriction on vesting wallet side.
Should we:
a. Remove restriction from here
b. Add restriction in wallet and duplicate restriction in factory
c. Add restriction in wallet and remove restriction in factory (since already check during initialization)
?

Reply:
I was referring the cliff being <= duration restriction. We just need to be sure that vesting wallet initialization won't revert as that would result in loss of funds. No need for any additional restrictions imo.

Reply:
Ok, see [Check cliff in batcher](https://github.com/OpenZeppelin/openzeppelin-confidential-contracts/pull/102/commits/064d19b71e2a0eb88bc13210805dc06fe50828ed).